materialize-elasticsearch/driver.go

```diff
@@ -304,19 +305,24 @@ func normalizeIndexName(index string, truncateLimit int) string {
 	afterPrefix := false
 	var b strings.Builder
 	for _, r := range index {
-		// Replace disallowed characters with underscore. Of the various problematic characters in
-		// an Elasticsearch index name, a "." is the only one that Flow allows in a collection name.
-		if r == '.' {
-			r = '_'
+		// Strip disallowed prefixes.
+		if !afterPrefix && slices.Contains([]rune{'_', '-', '+', '.'}, r) {
+			continue
 		}
 
-		// Strip disallowed prefixes that may be present in a Flow collection name. A '.' is also a
-		// bad prefix, but those were already replaced with '_' above.
-		if !afterPrefix && (r == '-' || r == '_') {
-			continue
+		// Replace disallowed characters with underscore as long as we aren't still at the beginning
+		// of the string. Most of the characters in this list are named in the docs, but ES will
+		// also truncate index names that contain a '#' character to drop everything including &
+		// after that '#' character, so we'll normalize those too.
+		if slices.Contains([]rune{'*', '<', '"', ' ', '\\', '/', ',', '|', '>', '?', ':', '#'}, r) {
+			if !afterPrefix {
```

Comment:
nit: is this on a separate line due to line length?

Reply:
It's a separate condition, since index names can't begin with an underscore and if a proposed index name has a disallowed character at the beginning it won't be valid to replace it with an underscore that would be at the beginning.

But looking at this now I think there's a simpler way to factor this that removes the second nested check of `if !afterPrefix` so I'll push that change.